README.md
Official website: http://libnss-mysql.sourceforge.net/
Buildstatus: [![Build Status](https://travis-ci.org/Cellebyte/libnss-mysql-bg-1.5.svg?branch=master)](https://travis-ci.org/Cellebyte/libnss-mysql-bg-1.5)

MySQL for NSS systems
=====================
Supported Operating Systems:
    o Linux
    o Solaris (Sparc or Intel) (2.6, 7, 8. 9 without PAM)
    o FreeBSD (5.1+)

Supported MySQL Versions:
    o MySQL 3.23.9 - 4.1.x

Supported Compilers:
    o GCC (2.95.2, 3.x)
    o Sun Forte Developer 7 C 5.4

NOTE: Solaris 9 is not fully unsupported.  Sun's PAM modules have become
      so restrictive that you can't even properly authenticate via PAM
      anymore if there's a "mysql" in /etc/nsswitch.conf.  System routines
      like getpwnam will still work, so you can work around it some by
      compiling your apps with PAM turned *off* - however logging in via
      system programs such as telnet will fail.  I'll be addressing this
      when I can; unfortunately it requires a PAM module that I'd need to
      write.  It's on my TODO list ...

Prerequisites
=============
    o Installing from source:
      o A functional compile environment (system headers, C compiler, ...)
      o MySQL client library & header files (local)
      o MySQL server (local or remote)

    o Installing from RPM:
      o MySQL server (local or remote)

    o Installing from Solaris Package
      o MySQL server (local or remote)

    o Installing from FreeBSD Port
      o MySQL client library (local)
      o MySQL server (local or remote)

The details
===========
    o If installing from source:
      o ./configure
      o make
      o make install
      If your MySQL installation is based in strange directory, use
      the --with-mysql=DIR option of ./configure to specify.  For example,
      "./configure --with-mysql=/usr2"

    o Edit /etc/libnss-mysql.cfg and /etc/libnss-mysql-root.cfg.

    o Add data to MySQL. The default configs will work well with the sample
      sql database in sample/sample_database.sql.  Read that file for more
      details on the sample database.

    o Edit (or create) /etc/nsswitch.conf such that it contains at least the
      following:
        passwd: files mysql
        shadow: files mysql
        group:  files mysql

      Do not enter the 'shadow' line on any system except Linux.
      If you don't want groups from MySQL, simply don't include 'mysql' in 
      in the 'group' line.

64-bit Support (SPARC V9/Solaris 7+)
====================================
You need to produce BOTH a 32-bit and 64-bit library, so get the 32-bit
one working first.  32-bit programs (most still are) will use the 32-bit
library, and 64-bit programs will use the 64-bit version.

You must have a set of 32-bit libraries, and another set of 64-bit
libraries: MySQL, libc, libm, etc ... Solaris tends to put 64-bit
libraries (and binaries) in a subdirectory named 'sparcv9' - so 64-bit
versions of libc, libm, etc, can be found in /usr/lib/sparcv9.

    o Build a 64-bit MySQL library and install it in a DIFFERENT place than
      your 32-bit MySQL library - IE /usr/local/lib/sparcv9/mysql
    o Build a 64-bit libnss-mysql:
      o Set CFLAGS settings in your environment to:
            gcc:      "-m64"
            Sun's CC: "-xarch=v9"
      o Re-run 'configure', pointing it at your 64-bit MySQL library
        (using --with-mysql=...)
      o Edit Makefile and change 'libdir' to be '/usr/lib/sparcv9'
      o Compile and install
    o Test it by using programs such as 'ls' and 'ps' from the /usr/bin/sparcv9
      directory.

At some point in the future, 64-bit support will be integrated.

Debugging
=========
    See the file DEBUGGING

$Id$


NEWS.md
# $Id$
April 10, 2017
    libnss-mysql 1.5-4 released

September 3, 2005
    libnss-mysql 1.5 released

January 30, 2005
    libnss-mysql 1.4 released

November 13, 2004
    libnss-mysql 1.3 released

March 28, 2004
    libnss-mysql 1.2 released

March 6, 2004
    libnss-mysql 1.1 released

July 12, 2003
    libnss-mysql 1.0 released

June 18, 2003
    libnss-mysql 0.9 released

January 25, 2003
    libnss-mysql 0.8 released

December 14, 2002
    libnss-mysql 0.7a released

December 4, 2002
    libnss-mysql 0.7 released

November 5, 2002
    libnss-mysql 0.6 released

October 23, 2002
    libnss-mysql 0.5 released

September 21, 2002
    libnss-mysql 0.4 released

August 22, 2002
    libnss-mysql 0.3 released

August 14, 2002
    libnss-mysql 0.2 released


src/lookup.c
/* Copyright (C) 2002 Ben Goodwin
   This file is part of the nss-mysql library.

   The nss-mysql library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as published
   by the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   The nss-mysql library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   You should have received a copy of the GNU General Public License
   along with the nss-mysql library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

static const char rcsid[] =
    "$Id$ ";

#include "nss_mysql.h"
#include <stdio.h>      /* snprintf () */
#include <string.h>     /* strcpy () */
#include <sys/types.h>
#include <grp.h>

extern conf_t conf;

/*
 * Take query from config, such as "SELECT x FROM foo WHERE bar = '%s'"
 * and replace "%s" with appropriate data
 * ltype = BYNAME, BYNUM, or BYNONE
 * name = username if this is a BYNAME
 * num = uid/gid if this is a BYNUM
 * qin = query in config to use
 * qout = string to send to MySQL server (after format string processing)
 * mresult = Current MySQL result (if any)
 * caller = calling function's name
 */
static NSS_STATUS
_nss_mysql_build_query (lookup_t ltype, const char *name, unsigned int num,
                        char *qin, char *qout, MYSQL_RES **mresult,
                        const char *caller)
{
  DN ("_nss_mysql_build_query")
  char clean_name[MAX_NAME_SIZE * 2 + 1];
  int retVal;

  DENTER

  /* Verify existence of input query, lest we crash */
  if (!qin || !qin[0])
    {
     _nss_mysql_log (LOG_CRIT, "%s has no valid query in config", caller);
     DSRETURN (NSS_UNAVAIL)
    }

  switch (ltype)
    {
    case BYNAME: /* This lookup key is a name/string */
      if (!name || !name[0])
        DSRETURN (NSS_NOTFOUND)
      D ("%s: BYNAME, name = '%s'", FUNCNAME, name);
      if (strlen (name) >= MAX_NAME_SIZE)
        {
          _nss_mysql_log (LOG_CRIT, "%s: name '%s' too long (MAX = %d)",
                          FUNCNAME, name, MAX_NAME_SIZE);
          DSRETURN (NSS_UNAVAIL)
        }
      if (_nss_mysql_escape_string (clean_name, name, mresult) != NSS_SUCCESS)
        DSRETURN (NSS_UNAVAIL)
      retVal = snprintf (qout, MAX_QUERY_SIZE, qin, clean_name);
      if (retVal < 1 || retVal >= MAX_QUERY_SIZE)
        {
          _nss_mysql_log (LOG_CRIT, "%s: snprintf error: %d", FUNCNAME, retVal);
          DSRETURN (NSS_UNAVAIL)
        }
      /* New lookup, reset any existing queries */
      _nss_mysql_reset_ent (mresult);
      break;
    case BYNUM: /* This lookup key is a uid/gid/number */
      D ("%s: BYNUM, num = '%u'", FUNCNAME, num);
      retVal = snprintf (qout, MAX_QUERY_SIZE, qin, num);
      if (retVal < 1 || retVal >= MAX_QUERY_SIZE)
        {
          _nss_mysql_log (LOG_CRIT, "%s: snprintf error: %d", FUNCNAME, retVal);
          DSRETURN (NSS_UNAVAIL)
        }
      /* New lookup, reset any existing queries */
      _nss_mysql_reset_ent (mresult);
      break;
    case BYNONE: /* This query has no key (e.g. a getpwent) */
      D ("%s: BYNONE creating initial query", FUNCNAME);
      strcpy (qout, qin);
      break;
    default:
      _nss_mysql_log (LOG_ERR, "%s: default case reached - should never happen",
                      FUNCNAME);
      DSRETURN (NSS_UNAVAIL)
    }
  DSRETURN (NSS_SUCCESS)
}

/*
 * The high-level function that does the actual lookup
 * Gets the query built and sent to the MySQL server, as well as
 * loads the result & buffer with the appropriate return data
 *
 * ltype = BYNAME, BYNUM, or BYNONE
 * name = name if BYNAME
 * num = uid/gid if BYNUM
 * q = unformatted query from config to use / build from
 * restricted = restrict this query to euid = 0?
 * result = NSS API result
 * buffer = NSS API buffer
 * buflen = NSS API length of buffer
 * errnop = NSS API errno pointer
 * load_func = pointer to function to load result/buffer with
 * mresult = current MySQL result set
 * caller = name of calling function
 */
NSS_STATUS
_nss_mysql_lookup (lookup_t ltype, const char *name, unsigned int num,
                   char *q, nboolean restricted, void *result,
                   char *buffer, size_t buflen, int *errnop,
                   NSS_STATUS (*load_func)(void *, char *, size_t,
                                           MYSQL_RES *, int *),
                   MYSQL_RES **mresult, const char *caller)
{
  DN ("_nss_mysql_lookup")
  char query[MAX_QUERY_SIZE];            /* Query to send to MySQL */
  int retVal;
  int attempts = MAX_QUERY_ATTEMPTS;    /* Attempt # (countdown) */
  static uid_t euid = -1;               /* Last known euid for change detect */
  uid_t cur_euid;                       /* CURRENT euid */
  gid_t cur_egid;                      /* CURRENT egid */
  gid_t shadow_gid;                    /* gid for group shadow (usually 42 on Debian) */

  DENTER

  cur_euid = geteuid ();

  /* Get shadow gid, if needed */
  if(cur_euid != 0) {
      cur_egid = getegid ();
      struct group *grp = getgrnam("shadow");
      shadow_gid = (grp ? grp->gr_gid : -1);
  }

  D ("%s: restricted = %d, cur_euid = %u", FUNCNAME, restricted, cur_euid);
  if (restricted == ntrue && cur_euid != 0 && (shadow_gid == -1 || cur_egid != shadow_gid))
    DSRETURN (NSS_NOTFOUND)

   /* Make sure euid hasn't changed, thus changing our access abilities */
  if (euid == -1)
    euid = cur_euid;
  else if (euid != cur_euid)
    {
      D ("%s:, euid changed: %u -> %u", FUNCNAME, euid, cur_euid);
      /* Close MySQL and config to force reload of config and reconnect */
      _nss_mysql_close_sql (mresult, ntrue);
      conf.valid = nfalse;
      euid = cur_euid;
    }

  /* Required in case link & config were reset due to euid change */
  if (_nss_mysql_init () != NSS_SUCCESS)
    DSRETURN (NSS_UNAVAIL)

  /* BYNONE indicates *ent; don't create/run the query since we already did */
  if (!(ltype == BYNONE && mresult && *mresult))
    {
      /* Create query string using config & args */
      retVal = _nss_mysql_build_query (ltype, name, num, q, query, mresult,
                                       caller);
      if (retVal != NSS_SUCCESS)
        DSRETURN (retVal)
      retVal = _nss_mysql_run_query (query, mresult, &attempts);
      if (retVal != NSS_SUCCESS)
        DSRETURN (retVal)
    }

  /* Take result of query and load RESULT & BUFFER */
  retVal = load_func (result, buffer, buflen, *mresult, errnop);

  /* BYNONE indicates *ent; don't kill the result here, endent does that */
  if (ltype != BYNONE)
    _nss_mysql_close_result (mresult);

  DSRETURN (retVal)
}


src/mysql.c
/* Copyright (C) 2002 Ben Goodwin
   This file is part of the nss-mysql library.

   The nss-mysql library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as published
   by the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   The nss-mysql library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with the nss-mysql library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * MySQL-specific functions; ALL MySQL calls that any other source functions
 * need should come from here.
 */

static const char rcsid[] =
    "$Id$";

#include "nss_mysql.h"
#include <string.h>     /* strlen() */
#include <netinet/in.h> /* struct sockaddr_in */

#ifndef HAVE_SOCKLEN_T
typedef size_t socklen_t;
#endif

con_info_t ci = { nfalse };
extern conf_t conf;

/*
 * Immediately after connecting to a MySQL server, save the current
 * socket information for later comparison purposes
 */
static freturn_t
_nss_mysql_save_socket_info (void)
{
  DN ("_nss_mysql_save_socket_info")
  socklen_t local_size = sizeof (struct sockaddr);
  socklen_t remote_size = sizeof (struct sockaddr);
  int r;

  DENTER
  memset(&(ci.sock_info.local), 0, sizeof (ci.sock_info.local));
  r = getsockname (ci.link.net.fd, &(ci.sock_info.local), &local_size);
  if (r != RETURN_SUCCESS)
    DFRETURN (RETURN_FAILURE)

  memset(&(ci.sock_info.remote), 0, sizeof (ci.sock_info.remote));
  r = getpeername (ci.link.net.fd, &(ci.sock_info.remote), &remote_size);
  DFRETURN (r)
}

/*
 * Compare ORIG and CUR
 */
static nboolean
_nss_mysql_is_same_sockaddr (struct sockaddr orig, struct sockaddr cur)
{
  DN ("_nss_mysql_is_same_sockaddr")

  DENTER
  switch (((struct sockaddr_in *)&ci.sock_info.local)->sin_family)
    {
    case AF_INET:
    case AF_INET6:
        if ((*(struct sockaddr_in *) &orig).sin_port !=
            (*(struct sockaddr_in *) &cur).sin_port)
          DBRETURN (nfalse)
        if ((*(struct sockaddr_in *) &orig).sin_addr.s_addr !=
            (*(struct sockaddr_in *) &cur).sin_addr.s_addr)
          DBRETURN (nfalse)
        break;
    case AF_UNIX:
        if (memcmp (&orig, &cur, sizeof (struct sockaddr)) != 0)
          DBRETURN (nfalse)
        break;
    default:
        _nss_mysql_log (LOG_ERR, "%s: Unhandled sin_family", FUNCNAME);
        DBRETURN (nfalse)
        break;
    }
  DBRETURN (ntrue)
}

/*
 * Check to see what current socket info is and compare to the saved
 * socket info (from _nss_mysql_save_socket_info() above).  Return
 * NTRUE if the current socket and saved socket match.
 */
static nboolean
_nss_mysql_validate_socket (void)
{
  DN ("_nss_mysql_validate_socket")
  socklen_t local_size = sizeof (struct sockaddr);
  socklen_t remote_size = sizeof (struct sockaddr);
  struct sockaddr check;

  DENTER
  memset(&check, 0, sizeof (check));
  if (getpeername (ci.link.net.fd, &check, &remote_size) != RETURN_SUCCESS)
    DBRETURN (nfalse)
  if (_nss_mysql_is_same_sockaddr (ci.sock_info.remote, check) != ntrue)
    DBRETURN (nfalse)

  memset(&check, 0, sizeof (check));
  if (getsockname (ci.link.net.fd, &check, &local_size) != RETURN_SUCCESS)
    DBRETURN (nfalse)
  if (_nss_mysql_is_same_sockaddr (ci.sock_info.local, check) != ntrue)
    DBRETURN (nfalse)

  DBRETURN (ntrue)
}

NSS_STATUS
_nss_mysql_close_sql (MYSQL_RES **mresult, nboolean graceful)
{
  DN ("_nss_mysql_close_sql")
  DENTER
  _nss_mysql_close_result (mresult);
  if (graceful && ci.valid)
    {
      D ("%s: calling mysql_close()", FUNCNAME);
      mysql_close (&ci.link);
    }
  ci.valid = nfalse;
  DSRETURN (NSS_SUCCESS)
}

static void
_nss_mysql_set_options (sql_server_t *server)
{
  DN ("_nss_mysql_set_options")
  const unsigned int def_timeout = DEF_TIMEOUT;
  const my_bool reconnect = 1;

  DENTER

  mysql_options(&ci.link, MYSQL_OPT_CONNECT_TIMEOUT,
                (const char *) &def_timeout);
  mysql_options(&ci.link, MYSQL_READ_DEFAULT_GROUP, "libnss-mysql");
#if MYSQL_VERSION_ID >= 50013
  mysql_options(&ci.link, MYSQL_OPT_RECONNECT, (const char *) &reconnect);
#else
  ci.link.reconnect = (my_bool) 1;
#endif

  DEXIT
}

/*
 * Validate existing connection.
 * This does NOT run mysql_ping because that function is
 * extraordinarily slow (~doubles time to fetch a query)
 */
static nboolean
_nss_mysql_check_existing_connection (MYSQL_RES **mresult)
{
  DN ("_nss_mysql_check_existing_connection")
  static pid_t pid = -1;

  DENTER
  if (ci.valid == nfalse || conf.valid == nfalse)
    DBRETURN (nfalse)

  if (pid == -1)
    pid = getpid ();
  else if (pid == getppid ())
    {
      /* saved pid == ppid = we've forked; We MUST create a new connection */
      D ("%s: fork() detected", FUNCNAME);
      ci.valid = nfalse;
      pid = getpid ();
      DBRETURN (nfalse)
    }

  if (_nss_mysql_validate_socket () == nfalse)
    {
       /* Do *NOT* CLOSE_LINK - the socket is invalid! */
      D ("%s: invalid socket detected", FUNCNAME);
      _nss_mysql_close_sql (mresult, nfalse);
      ci.valid = nfalse;
      DBRETURN (nfalse)
    }

  DBRETURN (ntrue)
}

/*
 * Connect to a MySQL server.
 */
static NSS_STATUS
_nss_mysql_connect_sql (MYSQL_RES **mresult)
{
  DN ("_nss_mysql_connect_sql")
  int retval;
  sql_server_t *server = &conf.sql.server;
  unsigned int port;

  DENTER

  if (_nss_mysql_check_existing_connection (mresult) == ntrue)
    {
      D ("%s: Using existing connection", FUNCNAME);
      DSRETURN (NSS_SUCCESS)
    }

  retval = _nss_mysql_load_config ();
  if (retval != NSS_SUCCESS)
    {
      _nss_mysql_log (LOG_ALERT, "Failed to load configuration");
      DSRETURN (NSS_UNAVAIL)
    }

  if (mysql_init (&ci.link) == NULL)
    {
      _nss_mysql_log (LOG_ALERT, "mysql_init() failed");
      DSRETURN (NSS_UNAVAIL)
    }

  _nss_mysql_set_options (server);
  D ("%s: Connecting to %s", FUNCNAME, server->host);
  if (server->port)
    port = atoi (server->port);
  else
    port = 0;
  if (mysql_real_connect (&ci.link, server->host,
                          server->username[0] ? server->username : NULL,
                          server->password[0] ? server->password : NULL,
                          server->database, port,
                          server->socket[0] ? server->socket : NULL,
                          0))
    {
      if (_nss_mysql_save_socket_info () != RETURN_SUCCESS )
        {
          _nss_mysql_log (LOG_ALERT, "Unable to save socket info");
          _nss_mysql_close_sql (mresult, ntrue);
          DSRETURN (NSS_UNAVAIL)
        }
      ci.valid = ntrue;
      ci.link.reconnect = 0; /* Safety: We can't let MySQL assume socket is
                                still valid; see _nss_mysql_validate_socket */
      DSRETURN (NSS_SUCCESS)
    }
  _nss_mysql_log (LOG_ALERT, "Connection to server '%s' failed: %s",
                  server->host, mysql_error (&ci.link));
  DSRETURN (NSS_UNAVAIL)
}

void
_nss_mysql_close_result (MYSQL_RES **mresult)
{
  DN ("_nss_mysql_close_result")
  DENTER
  if (mresult && *mresult && ci.valid)
    {
      D ("%s, calling mysql_free_result()", FUNCNAME);
      mysql_free_result (*mresult);
    }
  if (mresult)
    *mresult = NULL;
  DEXIT
}

/*
 * Run MySQL query
 */
NSS_STATUS
_nss_mysql_run_query (char *query, MYSQL_RES **mresult, int *attempts)
{
  DN ("_nss_mysql_run_query")
  int retval;

  DENTER
  if (!query)
    DSRETURN (NSS_NOTFOUND)

  D ("%s: Executing query: %s", FUNCNAME, query);

  retval = _nss_mysql_connect_sql (mresult);
  if (retval != NSS_SUCCESS)
    DSRETURN (NSS_UNAVAIL);

  retval = mysql_query (&ci.link, query);
  if (retval != RETURN_SUCCESS)
    {
      --(*attempts);
      if (*attempts > 0)
        {
          _nss_mysql_log (LOG_ALERT,
                          "mysql_query failed: %s, trying again (%d)",
                          mysql_error (&ci.link), *attempts);
          DSRETURN (_nss_mysql_run_query (query, mresult, attempts));
        }
      else
        {
          _nss_mysql_log (LOG_ALERT, "mysql_query failed: %s",
                          mysql_error (&ci.link));
          DSRETURN (NSS_UNAVAIL);
        }
    }

  if ((*mresult = mysql_store_result (&ci.link)) == NULL)
    {
      _nss_mysql_log (LOG_ALERT, "mysql_store_result failed: %s",
                      mysql_error (&ci.link));
      DSRETURN (NSS_UNAVAIL);
    }
  DSRETURN (NSS_SUCCESS)
}

NSS_STATUS
_nss_mysql_fetch_row (MYSQL_ROW *row, MYSQL_RES *mresult)
{
  DN ("_nss_mysql_fetch_row")

  DENTER
  if ((*row = mysql_fetch_row (mresult)) == NULL)
    {
      if (mysql_errno (&ci.link))
        {
          _nss_mysql_log (LOG_ALERT, "mysql_fetch_row() failed: %s",
                          mysql_error (&ci.link));
          DSRETURN (NSS_UNAVAIL)
        }
      else
        DSRETURN (NSS_NOTFOUND)
    }
  DSRETURN (NSS_SUCCESS)
}

NSS_STATUS
_nss_mysql_escape_string (char *to, const char *from, MYSQL_RES **mresult)
{
  DN ("_nss_mysql_escape_string")

  DENTER
  if (_nss_mysql_connect_sql (mresult) != NSS_SUCCESS)
    DSRETURN (NSS_UNAVAIL)
  mysql_real_escape_string (&ci.link, to, from, strlen(from));
  DSRETURN (NSS_SUCCESS)
}
